database/upload.go

```diff
@@ -80,10 +82,64 @@ func (db *DB) UploadsBySkylink(ctx context.Context, skylink Skylink, offset, pag
 	if err := validateOffsetPageSize(offset, pageSize); err != nil {
 		return nil, 0, err
 	}
-	matchStage := bson.D{{"$match", bson.D{{"skylink_id", skylink.ID}}}}
+	matchStage := bson.D{{"$match", bson.D{
+		{"skylink_id", skylink.ID},
+		{"unpinned", false},
+	}}}
 	return db.uploadsBy(ctx, matchStage, offset, pageSize)
 }
 
+// UploadsBySkylinkAndUser fetches all uploads of this skylink by this user.
+func (db *DB) UploadsBySkylinkAndUser(ctx context.Context, skylink Skylink, user User) ([]Upload, error) {
+	if skylink.ID.IsZero() {
+		return nil, errors.New("invalid skylink")
+	}
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	matchStage := bson.D{{"$match", bson.D{
+		{"skylink_id", skylink.ID},
+		{"user_id", user.ID},
+		{"unpinned", false},
+	}}}
+	c, err := db.staticUploads.Aggregate(ctx, mongo.Pipeline{matchStage})
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if errDef := c.Close(ctx); errDef != nil {
+			db.staticLogger.Traceln("Error on closing DB cursor.", errDef)
+		}
+	}()
+	var uploads []Upload
+	err = c.All(ctx, &uploads)
+	if err != nil {
+		return nil, err
+	}
+	return uploads, nil
+}
+
+// UnpinUploads unpins all uploads of this skylink by this user. Returns the
+// number of unpinned uploads.
```

Comment:
I wonder if I should limit this method to unpinning a single upload.

User story:  
As a user I upload a given skyfile twice by accident. The file is big and takes a long time to upload. I don't want it to take twice its size from my storage quota, so I unpin the second upload. I am surprised and frustrated when I notice (or even worse - not notice) that both my uploads are unpinned because due to the content-based hashing they have identical skylinks.

OTOH, it might be confusing to have to unpin the same file several times?

I'm not sure. I'm inclined to go with the first approach, as it makes more sense to me but I would like to get some feedback.

Reply:
We ended up with two endpoints:  
`DELETE /user/upload/:uploadId` which unpins a specific single upload  
`DELETE /skylink/:skylink` which unpins all.